llama.cpp

```diff
@@ -10003,151 +10024,41 @@ struct llm_tokenizer_bpe {
-                    (!token.size() && utf_char == " " && unicode_cpt_type(utf_char_next) != CODEPOINT_TYPE_LETTER && unicode_cpt_type(utf_char_next) != CODEPOINT_TYPE_DIGIT && unicode_cpt_type(utf_char_next) != CODEPOINT_TYPE_WHITESPACE)
-                    ) {
-                    collecting_special = true;
-                    collecting = true;
-                }
-                else if (unicode_cpt_type(utf_char) == CODEPOINT_TYPE_WHITESPACE && unicode_cpt_type(utf_char_next) == CODEPOINT_TYPE_WHITESPACE) {
-                    collecting_whitespace_lookahead = true;
-                    collecting = true;
-                }
-                else if (unicode_cpt_type(utf_char) == CODEPOINT_TYPE_WHITESPACE) {
-                    split_condition = true;
-                }
-            }
-            else if (!split_condition && collecting) {
-                if (collecting_letter && unicode_cpt_type(utf_char) != CODEPOINT_TYPE_LETTER) {
-                    split_condition = true;
-                }
-                else if (collecting_numeric && unicode_cpt_type(utf_char) != CODEPOINT_TYPE_DIGIT) {
-                    split_condition = true;
-                }
-                else if (collecting_special && (unicode_cpt_type(utf_char) == CODEPOINT_TYPE_LETTER || unicode_cpt_type(utf_char) == CODEPOINT_TYPE_DIGIT || unicode_cpt_type(utf_char) == CODEPOINT_TYPE_WHITESPACE)) {
-                    split_condition = true;
-                }
-                else if (collecting_whitespace_lookahead && (unicode_cpt_type(utf_char_next) == CODEPOINT_TYPE_LETTER || unicode_cpt_type(utf_char_next) == CODEPOINT_TYPE_DIGIT)) {
-                    split_condition = true;
-                }
-            }
-
-            if (utf_char_next == "") {
-                split_condition = true; // final
-                token += utf_char;
-            }
-
-            if (split_condition) {
-                if (token.size()) {
-                    bpe_words.emplace_back(token);
-                }
-                token = utf_char;
-                collecting = false;
-                collecting_letter = false;
-                collecting_numeric = false;
-                collecting_special = false;
-                collecting_whitespace_lookahead = false;
-            }
-            else {
-                token += utf_char;
-            }
-        }
-
-        for (std::string & word : bpe_words) {
-            std::string encoded_token = "";
-            for (char & c : word) {
-                encoded_token += unicode_byte_to_utf8(c);
-            }
-            bpe_encoded_words.emplace_back(encoded_token);
-        }
-
-        return bpe_encoded_words;
-    }
-
     const llama_vocab & vocab;
 
     std::vector<llm_symbol> symbols;
     std::vector<llm_symbol> symbols_final;
 
     llm_bigram_bpe::queue work_queue;
+
+    const std::vector<std::wstring> gpt2_regex = {
```

Comment:
I had a different idea about this - let me try to explain again:

In `llama.cpp`, we want to keep the original regex strings as they have been specified by the model creators. For example:

- `'s|'t|'re|'ve|'m|'ll|'d| ?\\p{L}+| ?\\p{N}+| ?[^\\s\\p{L}\\p{N}]+|\\s+(?!\\S)`
- `\s?\p{L}+`
- etc.

Now, my understanding is that in C++ we cannot simple perform some of those regex matching due to the lack of support for some the regex patterns in the standard library. So to solve this issue, we create the `unicode` module, which takes the regex strings from above as they are and performs a few different strategies to split the target string:

- If we have a known unicode representation generated in some way, we apply that using `std::wregex`. I.e. we check a constant `std::map<std::string, std::wstring>` for the presence of the regex
- If not, we then check if we have a custom implementation of the regex via a function call (see `bpe_gpt2_preprocess()` on master which is a custom implementation of regex split with `'s|'t|'re|'ve|'m|'ll|'d| ?\p{L}+| ?\p{N}+| ?[^\s\p{L}\p{N}]+|\s+(?!\S)|\s+`)
- Else, we just apply `std::regex` and hope for the best, or throw an error

Reply:
Hello, in the `unicode.cpp`, I have implemented the `unicode_regex_split()` function which iterates through the given regexes and if some match is found, it uses the regex. Otherwise, it  uses the modified `bpe_gpt2_preprocess()` function renamed as `unicode_custom_preprocess()`. Now, I have some questions regarding `bpe_gpt2_preprocess()`. Can it handle the input of deepseek coder and deepseek llm? If no, so, I have to write custom function for them when regex is not found?

Reply:
> Can it handle the input of deepseek coder and deepseek llm?

No, AFAIK it is not compatible with the deepseek regex. The way I understand it is that `bpe_gpt2_preprocess()` (i.e. `unicode_custom_preprocess()`) works only for the following regex (based on the comment in the code):

```
's|'t|'re|'ve|'m|'ll|'d| ?\p{L}+| ?\p{N}+| ?[^\s\p{L}\p{N}]+|\s+(?!\S)|\s+
```

So the logic in `unicode_regex_split()` has to check if this is the input regex and only apply this specific custom implementation in that case. For other regexes, we might want to implement more custom implementations in other functions and use them in `unicode_regex_split()` in the future.

Note that this my understanding of how this part of the tokenizer is supposed to work. I could be wrong, so don't take all of these suggestions for granted.

In any case, the huge unicode constants like `gpt2_regex` should not be located in `llama.cpp`, but instead should be in `unicode.cpp`.